From d53008720dff8cd1b3526424ab73a52e2f356701 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:04 -0400
Subject: [tools] perf/trace: Don't supress zeroed args when there is an strarray entry for it

Message-id: <1408462094-14194-136-git-send-email-jolsa@redhat.com>
Patchwork-id: 88087
O-Subject: [PATCH RHEL7.1 BZ1131394 135/325] perf trace: Don't supress zeroed args when there is an strarray entry for it
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 4aa5823225c70efba758750029504807cd30b925
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Fri Sep 20 12:19:41 2013 -0300

description
===========
Case in hand:

   9.682 ( 0.001 ms): Xorg/13079 setitimer(which: REAL, value: 0x7fffede42470) = 0

ITIMER_REAL is zero.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index 9f75fd0..4b30ec7 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -940,8 +940,15 @@ static size_t syscall__scnprintf_args(struct syscall *sc, char *bf, size_t size,
        field = field->next, ++arg.idx, bit <<= 1) {
    if (arg.mask & bit)
     continue;
-
-   if (args[arg.idx] == 0)
+   /*
+     * Suppress this argument if its value is zero and
+     * and we don't have a string associated in an
+     * strarray for it.
+     */
+   if (args[arg.idx] == 0 &&
+       !(sc->arg_scnprintf &&
+         sc->arg_scnprintf[arg.idx] == SCA_STRARRAY &&
+         sc->arg_parm[arg.idx]))
     continue;
 
    printed += scnprintf(bf + printed, size - printed,
-- 
1.7.1